overlord/devicestate/devicestate.go

```diff
@@ -293,6 +293,27 @@ func CanManageRefreshes(st *state.State) bool {
 	return false
 }
 
+// extractDownloadInstallEdgesFromTs is a helper that extract the first, last
+// download and install tasks from a TaskSet
+func extractDownloadInstallEdgesFromTs(ts *state.TaskSet) (firstDl, lastDl, firstInst, lastInst *state.Task) {
+	edgeTask := ts.Edge(snapstate.DownloadAndChecksDoneEdge)
+	for _, t := range ts.Tasks() {
+		if firstDl == nil {
+			firstDl = t
+		}
+		if firstInst == nil && lastDl != nil {
+			firstInst = t
+		}
+		if t == edgeTask {
+			lastDl = t
+		}
+		if firstInst != nil {
+			lastInst = t
+		}
```

Comment:
firstDl and lastInst  are just  tasks[0] and tasks[-1], no?  it seems slightly silly to use the loop for that

the loop is only needed to find firstInst, which comes after lastDl (which is just edgeTask)

(I still have a  slight preference for doInstall to mark all these but it could be done later, it's not a blocker per se)